src/v/kafka/client/consumer.cc

```diff
@@ -133,7 +133,7 @@ ss::future<> consumer::stop() {
     _inactive_timer.cancel();
     _inactive_timer.set_callback([]() {});
 
-    _on_stopped(_name);
+    _on_stopped(name());
```

Comment:
i'm having trouble understanding part of the logic for the fix.

> in the case of an unnamed consumer because `_name`

are unnamed consumers valid?

>     _consumers[group_id].erase(name); // silenty fails if name is empty

if unnamed consumers are valid, then why would this fail to erase it? i wouldn't think that erase would care if it were an empty nor non-empty string.

> consumer::name()

what does this do differently that solves the problem?

Reply:
The user can optionally name a consumer. If a name is provided, that's the key in the map, otherwise it is looked up by the `member_id` provided by Redpanda (since that's returned to the user if a name isn't provided)

Reply:
> The user can optionally name a consumer. If a name is provided, that's the key in the map, otherwise it is looked up by the `member_id` provided by Redpanda (since that's returned to the user if a name isn't provided)

@dotnwat  To add to this explanation ....


> > in the case of an unnamed consumer because `_name`
> 
> are unnamed consumers valid?

Yes, they are valid. A user could optinally supply a name in the json. 

> > ```
> > _consumers[group_id].erase(name); // silenty fails if name is empty
> > ```
> 
> if unnamed consumers are valid, then why would this fail to erase it? i wouldn't think that erase would care if it were an empty nor non-empty string.

You are correct in that erase does **not** care is the string is empty. However, when the consumer is created, it is inserted into the map based on the result of `consumer::name()` and `consumer::name()` will return a non-empty string in the case of an unnamed consumer. 

Therefore, a call to `_consumers[group_id].erase(name);` where name is an empty string is actually an attempt to erase the a different consumer hashed by an empty string. Furthermore, the particular overload of erase that is used here does not throw if an element is not found. Instead, it returns a size_t representing the number of elements deleted and because no consumer was hashed with an empty string, this call will return 0 (See the 4th def of unordered map [here](https://en.cppreference.com/w/cpp/container/unordered_map/erase)).

TLDR - we were attempting to erase the consumer identified by a hash on an empty string instead of the consumer identified by a hash of `consumer::name()`.

> 
> > consumer::name()
> 
> what does this do differently that solves the problem?

Here is the definition of `consumer::name()`:
```cpp
    const kafka::member_id& name() const {
        return _name != kafka::no_member ? _name : _member_id;
    }
```
When a consumer is unnamed, `consumer::name()` will return the member id (a non-empty string). So to re-iterate:
1. On creation of an unnamed consumer, we insert it into a map based on the member_id returned from `consumer::name()`:
```cpp
struct consumer_hash {
    // .. some stuff ..
    size_t operator()(const shared_consumer_t& c) const {
        return (*this)(c->name()); // Our hasher calls consumer::name()
    }
};

.then([this, group_id](shared_consumer_t c) {
   auto name = c->name();
   // We insert into map based on consumer ptr but the hasher will actually use the result of consumer::name()
   _consumers[group_id].insert(std::move(c));
   return name;
});
```
2.  **But** we were erasing based on the `_name` var ehich is an empty-string, this is an entirely different key.
```cpp
_on_stopped(_name); // _name is empty string

auto on_stopped = [this, group_id](const member_id& name) {
    // Attempt to erase with an empty string as the key
    _consumers[group_id].erase(name);
};
```

By using `consumer::name()` as the argument to `on_stopped`, we ensure that the correct string is used as the key.

Reply:
> ahh i see so name() encodes those rules?

Yes. (I should've seen your comment before posting my long one :) )

Reply:
thanks @NyaliaLui that's a great explanation!